atc/lidar/scanner.go

```diff
@@ -60,7 +60,7 @@ func (s *scanner) Run(ctx context.Context) error {
 		go func(resource db.Resource, resourceTypes db.ResourceTypes) {
 			defer waitGroup.Done()
 
-			err = s.check(resource, resourceTypes)
+			err := s.check(resource, resourceTypes)
 			s.setCheckError(s.logger, resource, err)
```

Comment:
so access to the `err` variable was not thread-safe, but is `setCheckError`? I guess so, because we assume the database will synchronize across concurrent queries, but I'm not totally sure.

Reply:
@pivotal-jamie-klassen The problem is, when multiple go routines operates the same object `err`, there is a race condition. For example, in go routine 1, `check` returns an error, then `setCheckError` is called. Right when `err.Error()` is called in routing 1, routine 2 may set `err` to nil, then in routine 1, `err.Error()` will crash.

Reply:
From what I can see you still have a data race, though, with multiple goroutines sharing unsynchronized access to the scanner `s` here. In the scenario you're describing, this race gets pushed down to the database, so the `check_error` that gets written in the database will depend on which goroutine does the last write, or maybe which one gets the lock on that row, depending on how Postgres synchronization works (I'm not really sure). What I'm wondering is, have you thought this through? Should we be careful about what ends up getting written in the database, or are you convinced that we have sufficient synchronization?

Reply:
@pivotal-jamie-klassen `s.setCheckError()` is stateless, it just call stores the error to database, so there should be no unsynchronized access problem for `s` here.

Reply:
(a side thought here): if this is the kind of thing that `-race` could catch - it'd probably be in our favor to have either smoke tests or topgun ones (that are "end-to-end" in nature) running with a `-race` bug 🤔

Reply:
@cirocosta I agree. But for this kind of bug, it's really not easy to trigger, so I'm afraid a unit or topgun test case would be hard to catch this bug.